### ISSUE TITLE
Consider adding convenience workaround so sanitizer builds of Zeek can be run on macOS

### ISSUE THREAD

**Issue by bbannier** [zeek/zeek#3347]
The default clang compiler which ships with macOS does not support leak sanitizer. This means that users who want to run a sanitizer build of Zeek need to explicitly set `ASAN_OPTIONS=detect_leaks=0` in their environment.

We could consider making this more streamlined. For developers running out of the build directory we could inject above `ASAN_OPTIONS` if the developer environment doesn't already define `ASAN_OPTIONS`. For this to work for installations we would need to embed changed ASAN defaults into all binaries we install (since `-fsanitze=address` is viral), e.g., add the following to all binaries we ship once (ODR etc.):

```cpp
#ifdef __cplusplus
extern "C"
#endif
const char* __asan_default_options() { return "detect_leaks=0"; }
```

Until this is ship, anyone wishing to run binaries build with ASAN on macOS with a compiler which does not support LSAN should set in their environment (probably statically since the default behavior of clang is just broken there)

```console
export ASAN_OPTIONS=detect_leaks=0
```

**Comment by awelzel** [comment]
> For this to work for installed binaries we would need to embed changed ASAN defaults into all binaries we install (since -fsanitze=address is viral), e.g., [add the following](https://github.com/google/sanitizers/wiki/AddressSanitizerFlags#run-time-flags) to all binaries we ship exactly once (ODR etc.):

But only for OSX, right?


> The default clang compiler which ships with macOS does not support leak sanitizer. This means that users who want to run a sanitizer build of Zeek need to explicitly set ASAN_OPTIONS=detect_leaks=0 in their environment.

I'm not fully following #3026. It reads as if @vpax  gets errors about unsupported LSAN when running `zeek` when building with address sanitizer and Spicy enabled, but then the errors go away when building with `--disable-spicy`. Or did I miss something here?

---

Mentioning the following bifcl/binpac hunks below as semi-related in that for those executables LSAN is disabled explicitly, otherwise they'd likely cause non-functional builds on OSX today, too. Background in #699.

The `__asan_default_options()` approach seems much nicer, though.

https://github.com/zeek/bifcl/blob/c7bf54c587439d3bcb16d53b0d77a702e48d2526/builtin-func.l#L319-L325

https://github.com/zeek/binpac/blob/84b730fdcc5b983c65c6226ec092aee66c486680/src/pac_main.cc#L219-L232

**Comment by timwoj** [comment]
> We actually have the LSAN macro games also in Zeek, but we do enable LSAN in the end from what it seems.

Correct, the block in `zeek-config.h.in` only disables it if we're on FreeBSD. I'm a bit torn on globally disabling it as @bbannier's block would do. I'd prefer to change the macros in zeek-config.h/bifcl/binpac to correctly disable it on macOS as well as FreeBSD.

**Comment by timwoj** [comment]
I'm wavering on the right path here. I don't really like the idea of defaulting the flags in Zeek via `__asan_default_options`, because later setting them `ASAN_OPTIONS` overrides whatever the defaults are. Checking `__APPLE__` in `zeek-config.h.in` fixes at least the LSan problem in this ticket, but as https://github.com/zeek/zeek/issues/4077 shows we could also be defaulting the `detect_container_overflow` and `detect_odr_violation` to off.

**Comment by vpax** [comment]
I may not have followed this sufficiently, but FWIW `ASAN_OPTIONS=detect_leaks=0` is no longer needed under MacOS. What _is_ needed is `detect_container_overflow=0` since otherwise that gets tickled on any startup (in Broker code IIRC).
